File: csharp/interest-is-interesting/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
    public static float InterestRate(decimal balance) => balance switch { < 0 => -3.213f, < 1000 => 0.5f, < 5000 => 1.621f, _ => 2.475f };

    public static decimal Interest(decimal balance)
    {
        decimal interestRatePercentage = (decimal)(InterestRate(balance) / 100);
        
        return interestRatePercentage * Math.Abs(balance);
    }

    public static decimal AnnualBalanceUpdate(decimal balance) => balance + Interest(balance);

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        int yearsOfInterest = 0;

        do
        {
            balance = AnnualBalanceUpdate(balance);
            yearsOfInterest++;
        } while (balance < targetBalance);

        return yearsOfInterest;
    }
}

File: fsharp/interest-is-interesting/InterestIsInteresting.fs
module InterestIsInteresting

let interestRate (balance: decimal): single =
    if balance < 0m then
        3.213f
    elif balance < 1000m then
        0.5f
    elif balance < 5000m then
        1.621f
    else
        2.475f

let interest (balance: decimal): decimal = balance * (decimal (interestRate balance) / 100m)

let annualBalanceUpdate(balance: decimal): decimal = balance + interest balance

let amountToDonate(balance: decimal) (taxFreePercentage: float): int =
    if balance > 0.0m then
        int (balance * decimal (taxFreePercentage / 100.0) * 2m)
    else
        0
